packages/worker/src/worker.ts

```diff
@@ -993,260 +998,261 @@ export class Worker {
+      filter((state) => state !== 'POLLING'),
+      first(),
+      map(() => ({
+        activation: coresdk.workflow_activation.WorkflowActivation.create({
+          runId: activations$.key,
+          jobs: [{ removeFromCache: Worker.SELF_INDUCED_SHUTDOWN_EVICTION }],
+        }),
+        synthetic: true,
+      })),
+      takeUntil(activations$.pipe(last(undefined, null)))
+    );
+    const activations$$ = activations$.pipe(map((activation) => ({ activation, synthetic: false })));
+    return merge(activations$$, evictions$).pipe(
+      tap(() => {
+        this.numInFlightActivationsSubject.next(this.numInFlightActivationsSubject.value + 1);
+      }),
+      mergeMapWithState(this.handleActivation.bind(this), undefined),
+      tap(({ close }) => {
+        this.numInFlightActivationsSubject.next(this.numInFlightActivationsSubject.value - 1);
+        if (close) {
+          activations$.close();
+          this.numCachedWorkflowsSubject.next(this.numCachedWorkflowsSubject.value - 1);
+        }
+      }),
+      takeWhile(({ close }) => !close, true /* inclusive */)
+    );
+  }
+
+  /**
+   * Process a single activation to a completion.
+   */
+  protected async handleActivation(
+    state: WorkflowWithLogAttributes | undefined,
+    { activation, synthetic }: { activation: coresdk.workflow_activation.WorkflowActivation; synthetic: boolean }
+  ): Promise<{
+    state: WorkflowWithLogAttributes | undefined;
+    output: { completion?: Uint8Array; close: boolean };
+  }> {
+    try {
+      const removeFromCacheIx = activation.jobs.findIndex(({ removeFromCache }) => removeFromCache);
+      const close = removeFromCacheIx !== -1;
+      const jobs = activation.jobs;
+      if (close) {
+        const asEvictJob = jobs.splice(removeFromCacheIx, 1)[0].removeFromCache;
+        if (asEvictJob) {
+          this.evictionsEmitter.emit('eviction', {
+            runId: activation.runId,
+            evictJob: asEvictJob,
+          } as EvictionWithRunID);
+        }
+      }
+      activation.jobs = jobs;
+      if (jobs.length === 0) {
+        this.log.trace('Disposing workflow', {
+          ...(state ? state.logAttributes : { runId: activation.runId }),
+        });
+        await state?.workflow.dispose();
+        if (!close) {
+          throw new IllegalStateError('Got a Workflow activation with no jobs');
+        }
+        const completion = synthetic
+          ? undefined
+          : coresdk.workflow_completion.WorkflowActivationCompletion.encodeDelimited({
+              runId: activation.runId,
+              successful: {},
+            }).finish();
+        return { state: undefined, output: { close, completion } };
+      }
+
+      if (state === undefined) {
+        // Find a workflow start job in the activation jobs list
+        const maybeStartWorkflow = activation.jobs.find((j) => j.startWorkflow);
+        if (maybeStartWorkflow !== undefined) {
```

Comment:
Add the following to the top of this file (next to the `PatchJob` type):

```
type StartWorkflowJob = NonNullableObject<Pick<coresdk.workflow_activation.IWorkflowActivationJob, 'startWorkflow'>>;
```

then modify this line to:

```
        const maybeStartWorkflow = activation.jobs.find((j): j is StartWorkflowJob => j.startWorkflow != null);
```

By doing this, when you get you will know that `maybeStartWorkflow.startWorkflow` is defined.

Reply:
Thanks! Type guards are one of several things I've learned about Typescript from working on this codebase.  Maybe double check that https://github.com/temporalio/sdk-typescript/commit/3c74d2a044f5642dbb1bdedfa098a1cf4c2eaaed does what you intended.